### PR TITLE
[enough to visualize a-star] simulator

### PR DESCRIPTION
**Pull request by s4ayub**
**to do**:

- [x] test travelling from point to point (write world.travel( and robot.travel( functions 
- [x] figure out how to draw our a_star progression on robot's version of the world
- [x] write a_star algorithm
- [ ] try out different heuristic scaling factors (refer to: http://theory.stanford.edu/~amitp/GameProgramming/Heuristics.html) (0.8 multiplier for adding turn, determine what a turn is, 1.2 multiplier for unknown tile, etc. etc.)
- [ ] convert to cpp

**notes**:
say you have a path like: [[0, 1] [0,2], [0,3]], **but**  upon detection, [0, 1] is a water tile. 
You shouldn't call a_star again with [0, 3] as the target, rather you should call it with [0, 2] as the target, since it's closer and it'll take less time to process, then you can prepend the coordinates generated to your original path. 

**ACTUALLLYYYYY** discussed this with @arnettus, better to just do a_star from new position due to water block to end because there might be new more efficient path. Think about possible caching strategies to make this a_star faster? because we're repeating a lot of work. Although, we probably dont have the memory for that.

clean-up:
- [x] turn world into a class
- [x] throw constants in main.py into constants.py, import into every file
- [ ] use __main__.py python convention something something

### CODE REVIEW

**Comment by s4ayub** [comment]
right nowwww:
- make sure you can actually move the mudkip
- then, repalce all rectangles with 60px colored boxes, easier to move images than to move rectangles

**Comment by s4ayub** [comment]
right nowwww:
- [ ] write the algorithm with no costs or tiles or whatever, just in a separate place, then test it to go from point a to point b, like straight line, bring it into the main code,
- [ ] add detect tile functionality to dodge water, update tiles as you see them
- [ ] adjust the heuristc and g-cost to take into account the different types of tiles,

**Comment by s4ayub** [comment]
Note:
- [ ] find PQ in c++

**Comment by s4ayub** [comment]
show frontierrrrrrr
---
the robot's map should store the state of what it has seen
robot's pos stores what it has seen

show the thinking, 

clear screen

update screen with robot's map

File: software/kowloon/path.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+def heuristic(a, b):
+    (x1, y1) = a
+    (x2, y2) = b
+    return abs(x1 - x2) + abs(y1 - y2)
+
+# this is the weight of the edge???
+def cost(self, from_node, to_node):
+    # adjust this based on tile costttt
+    return 1
```
[the remark this answers is not in the document]
Comment (DrynnBavis) [reply software/kowloon/path.py]: Ye I like that idea. One other (though probably worse idea) would be to ignore heuristic values when pathing until the food is located. Essentially using dijkstra for more of a BFT, maximising the tiles visited, until the food is found. Then changing the weights in our stack to include the heuristics (returning back to A-star) and re-sorting.

[the remark this answers is not in the document]
Comment (s4ayub) [reply software/kowloon/path.py]: Djisktra and A-star try to achieve the same thing, find shortest path, djisktra is just slower at doing it. 

I was planning to add diff parameters to the heuristic to achieve what you're talking about. Where, before the food is found, the heuristic makes it so you don't necessarily find the shortest path, but you find a happen medium between a short path and a path with a lot of unknowns.

afterwards, all that matters is the shortest path.

**Pull request merged**